File: ocaml/apps.ml
let re_app_name = Str.regexp "^[^./\\\\:=;'\"][^/\\\\:=;'\"]*$";;

let lookup_app name config =
  if Str.string_match re_app_name name 0 then
    Basedir.load_first_config ["0install.net"; "apps"; name] config.Config.basedirs
  else
    None
;;

File: haskell/Apps.hs
module Apps where

import Text.Regex.Posix
import System.FilePath

import Config
import Basedir

validAppName :: String -> Bool
validAppName name = name =~ "^[^./\\\\:=;'\"][^/\\\\:=;'\"]*$"

lookupApp :: String -> Config -> IO (Maybe FilePath)
lookupApp name config | validAppName name = loadFirst ("0install.net" </> "apps" </> name) (Basedir.config $ basedirs config)
		       | True = return Nothing
